sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample11_MockingClientTypes.md

```diff
@@ -0,0 +1,502 @@
+ Mocked batches accept a `List<EventData>` that is used as a backing store and can be inspected to verify that the application is adding events to the batch as expected.  The custom `TryAdd` callback can be used to control the decision for whether an event is accepted into the batch or is rejected.   
+
+This snippet demonstrates mocking the `EventHubProducerClient` using Moq, and creating `EventDataBatch` instances using the `EventHubsModelFactory`.
+
+```C# Snippet:EventHubs_Sample11_MockingEventDataBatch
+Mock<EventHubProducerClient> mockProducer = new();
+
+// This sets the value returned by the EventDataBatch when accessing the Size property,
+// it does not impact TryAdd on the mocked batch.
+
+long batchSizeInBytes = 500;
+
+// Events added to the batch will be added here, but altering the events in this list will not change the
+// events in the batch, since they are stored inside the batch as well.
+
+List<EventData> backingList = new();
+
+// For illustrative purposes, this is the number of events that the batch will contain,
+// returning false from TryAdd for any additional calls.
+
+int batchCountThreshold = 3;
+
+EventDataBatch dataBatchMock = EventHubsModelFactory.EventDataBatch(
+        batchSizeBytes : batchSizeInBytes,
+        batchEventStore : backingList,
+        batchOptions : new CreateBatchOptions(),
+        // The model factory allows the user to define a custom TryAdd callback, allowing
+        // control of what events the batch accepts.
+        eventData =>
+        {
+            int eventCount = backingList.Count();
+            return eventCount < batchCountThreshold;
+        });
+
+// This sets up a mock of the CreateBatchAsync method.
+
+mockProducer
+    .Setup(p => p.CreateBatchAsync(It.IsAny<CancellationToken>()))
+    .ReturnsAsync(dataBatchMock);
+
+// Here we are mocking the SendAsync method so that it will throw an exception if the batch passed
+// into send is not the one we are expecting to send.
+
+mockProducer
+    .Setup(p => p.SendAsync(
+        It.Is<EventDataBatch>(sendBatch => sendBatch != dataBatchMock),
+        It.IsAny<CancellationToken>()))
+    .Throws(new Exception("The batch published was not the expected batch."));
+
+EventHubProducerClient producer = mockProducer.Object;
+
+EventDataBatch batch = await producer.CreateBatchAsync();
+
+// Next we are creating a list of events to use in our test.
+
+List<EventData> sourceEvents = new();
+
+for (int index = 0; index < batchCountThreshold; index++)
+{
+    var eventData = new EventData($"Sample-Event-{ index }");
+    sourceEvents.Add(eventData);
+}
+
+// Here we are adding events to the batch. They should all be accepted.
+
+foreach (var eventData in sourceEvents)
+{
+    Assert.True(batch.TryAdd(eventData));
+}
+
+EventData eventData4 = new EventData(eventBody: new BinaryData("Sample-Event-4-will-fail"));
```

Comment:
We probably want a comment above this mentioning that since have `batchCountThreshold` events in there, the batch is considered full by the callback that we passed.   The next event should be rejected.